x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/predicate/operator/arithmetic/DateMathException.java

```diff
@@ -0,0 +1,33 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.esql.expression.predicate.operator.arithmetic;
+
+import org.elasticsearch.rest.RestStatus;
+import org.elasticsearch.xpack.esql.EsqlClientException;
+import org.elasticsearch.xpack.ql.tree.Source;
+
+/**
+ * Represents e.g. overflows when folding temporal values caused by a faulty query.
+ * This is turned into a REST response with HTTP status (400).
+ * Even though this is a pretty specific exception, we cannot use e.g. a anonymous class that inherits from {@link EsqlClientException},
+ * because then the error message sent back in the REST response would have an empty `type` string.
+ */
+public class DateMathException extends EsqlClientException {
```

Comment:
The users will not see the exception just the message thus there's no value in having a different exception type.
How many places are going to throw this error?
If there are multiple where the same error is thrown, it makes to centralize this into a class that creates the same EsqlClientException and return different messages.

Reply:
Unfortunately, this is not correct. The name of the exception is user-facing because of how the exception handling of our action listener framework works. When the REST response is being built in case of an error, the name of the exception class becomes the "type"; in the current state of this PR, the following query
```
row n = now() | eval n + (2147483647 year + 1 year)
```
results in the following (complete) response
```
{"error":{"root_cause":[{"type":"date_math_exception","reason":"arithmetic exception in expression [2147483647 year + 1 year]: [integer overflow]"}],"type":"date_math_exception","reason":"arithmetic exception in expression [2147483647 year + 1 year]: [integer overflow]"},"status":400}
```
Note the `"type": "date_math_exception"`. This was generated automatically, I don't have control over this.

Even though there's currently only 2 places where this exception is needed, it still kinda has to exist because the exception should inherit from `EsqlClientException` (abstract) and it should not be anonymous because of the reason above.

Since exceptions during constant folding is a general problem (see https://github.com/elastic/elasticsearch/issues/99758), I suspect that this exception can later be refactored into something more general, like `ConstantFoldingException`. I didn't use a more general name because names are hard and that discussion could be deferred until fold-time exceptions are handled consistently. But if you'd like me to create a more generic exception that inherits from `EsqlClientException`, instead, I'm happy to do so - maybe `ExpressionException`? (`VerificationException` comes to mind but is not really precise.)

Reply:
Exception hierarchies in Java are tricky because they're abuse while they're suppose to have meaning.
It's the error message that is relevant, not the exception type. Since `EsqlClientException` is abstract you can use `ParsingException` or `VerificationException` - neither is really appropriate but will do for now.